text/0085-api-keys.md

```diff
@@ -0,0 +1,550 @@
+
+If the master key is removed at MeiliSearch launch, the previously generated API keys no longer secure the MeiliSearch instance.
+
+If MeiliSearch is launched with the `production` value for the `MEILI_ENV` environment variable or the `--env` CLI option, a master key is mandatory to force the user to secure his instance. If the master key is omitted in that particular case, MeiliSearch launch is aborted and displays the `Error: In production mode, the environment variable MEILI_MASTER_KEY is mandatory` error in stdout.
+
+The master key must be composed of valid utf-8 characters. It is advisable to enclose it in `'` when specified via the --master-key option.
+
+> 🚨 A user coming from a version prior to v0.25.0 and having a master-key will have to update their code to use the newly generated default keys replacing the public and private, since we changed the previous `public` and `private` API Key generation, is it mandatory for that specific version upgrade. Now, a prefix of an API Key is generated uniquely and the final value of the `key` field is a hash of that randomized prefix with the master-key. See 2.1 API Key Generation part.
+
+> 🚨 The master-key should never be exposed to the public or bad-intentioned persons for security measures as it may compromise a MeiliSearch instance.
+
+> 🚨 If the value of the master key changes, all the previously generated `API Keys` changes, thus allows to invalidate the set of keys previously generated by regenerating to a different value for the `key` field. This is particularly useful in the case where the master key might have been leaked and the user need to re-generate the whole set of keys at once to re-secure the instance.
+
+> Note that the master key does not appear on the `/keys` endpoints.
+
+#### 1.5.3 Default API Keys
+
+When the user accessing the machine launches MeiliSearch with a `master` key the first time, MeiliSearch will generate two API keys described below, as it did before with the `public` and `private` key.
+
+If the user changes the value of the `master` key later, these two default keys are not created again but re-generated with a different `key` field. However, these two API keys can be changed using the `/keys` endpoints.
+
+MeiliSearch must know that it has already generated these Default API Keys internally so if the user delete them, the engine should not create them again when MeiliSearch is launched again with a `master` key.
+
+##### 1.5.3.1 Default Search API Key
+
+The `Default Search API key` gives access to the same rights as the old `public` key.
+
+Here is how the `Default Search API Key` is represented after its generation.
+
+```json
+{
+    "description": "Default Search API Key (Use it to search from the frontend)",
```

Comment:
There is any possibility to prevent the user from using the wrong key in the frontend, and making their infrastructure unsecured?

Maybe an alert, log, or something else in the mini-dashboard. 🌳

Reply:
We can't prevent the use of a "bad" key, because we can't know where it is called. We have added the description to help as much as possible a MeiliSearch user to use the right key for the right purposes. Adding logs could be very verbose but it can be considered according to the feedback and the use that will be made of it.

Reply:
For what it's worth, we do our best in the docs to make it clear to users they need to be careful with keys—especially when it comes to the master key.

Which reminds me of something @gmourier and I spoke about a while ago. The master key can access all API endpoints, even when by design we want people to only use it when managing API keys. We could go a step further beyond warnings and simply block the master key specifically from accessing any routes except `/keys`. This wouldn't have an effect on API keys, though.

Reply:
> We could go a step further beyond warnings and simply block the master key specifically from accessing any routes except /keys. This wouldn't have an effect on API keys, though.

Well, I think that definitely will prevent this kind of mistake @guimachiavelli, (I'm always a pessimist on these subjects 😅)

Reply:
@guimachiavelli That's a good idea, subject to debate because the general philosophy that seems to prevail is not to restrict or force the hand of the users by blocking him. We had/have many debates on this subject with @bidoubiwa during our design reviews/feedbacks discussion, that said, from a purely personal point of view when it comes to security, I'm rather for restricting in order to better accompany the user rather than leaving him the possibility to put himself in danger.

Reply:
@gmourier: apologies if it sounded as if I were saying that we should totally do it 😅. Like you and @brunoocasali, I'm also for restricting harmful behaviour, but there's certainly a case to be made for educating rather than forcing users to "do the right thing."